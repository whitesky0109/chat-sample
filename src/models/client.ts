import { UserPayload } from 'public/store/actions/userAction';
import { RoomPayload } from 'public/store/actions/roomAction';

export interface IMessage {
  roomId?: string;
  type: string;
  message: File | string;
}

export interface StoreState {
  user: UserPayload;
  room: RoomPayload;
}
