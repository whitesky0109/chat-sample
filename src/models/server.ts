export interface IMigration {
  migration_id: string;
  action: string;
  args: any[];
}

export interface IRepository {
  migrations(): IMigration[];
}

export interface IService {
  init(): Promise<any>;
}

export interface IMigrationLog {
  id: number;
  migration_id: string;
  action: string;
  success: boolean;
  error: string;
  timestamp: Date;
}

export interface IMigrationLogs {
  // tslint:disable-next-line: variable-name
  [migration_id: string]: IMigrationLog;
}

export interface IInvite {
  roomId: string;
  users: string[];
}

export interface IMessage {
  userId: string;
  type: string;
  message: ArrayBuffer | string;
  date: Date;
  isMy: boolean;
}

export interface IUser {
  [userId: string]: string | null;
}

export interface IRoomInfo {
  name: string;
  users: IUser;
  creator: string;
}

export interface IRoom {
  [roomId: string]: IRoomInfo;
}
