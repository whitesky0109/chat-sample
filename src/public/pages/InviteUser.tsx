import _ from 'lodash';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import React, { Component, ReactNode } from 'react';

// models
import { StoreState } from 'models/client';
import { IUser, IInvite, IRoom } from 'models/server';

// etc
import { reqUsers, reqInvite } from '../controllers/socket';

export interface Props extends RouteComponentProps {
  room?: IRoom;
  roomId?: string;
  user?: string;
  users?: IUser;
}

export class InviteUser extends Component<Props> {

  componentWillMount() {
    const { roomId, history } = this.props;

    if (!roomId) {
      history.push('/room');
      return;
    }

    reqUsers();
  }

  shouldComponentUpdate(nextProps: Props) {
    const { history } = this.props;

    if (!nextProps.roomId) {
      history.push('/room');
      return false;
    }

    return true;
  }

  onInviteUser = (id: string) => {
    const { roomId, history } = this.props;
    if (!roomId) {
      return;
    }

    const msg: IInvite = {
      roomId,
      users: [id],
    };

    reqInvite(msg);
    history.push(`/room/${roomId}`);
  }

  renderList(): ReactNode[] {
    const { users } = this.props;

    if (!users) {
      return [];
    }

    const result: ReactNode[] = _.map(users, (user, id: string) => {
      const { room, roomId } = this.props;

      if (!room || !roomId) {
        return;
      }
      const { users: invitedUsers } = room[roomId];
      const userIds = _.map(invitedUsers, (_, userId) => userId);

      if (userIds.includes(id)) {
        return;
      }

      return (<li className="user clearfix" key={id} onClick={this.onInviteUser.bind(this, id)}>
        <div className="user-icon">
          <i className="fa fa-users" />
        </div>
        <div className="about">
          <div className="name">{id}</div>
          <div className="status">
            <i className="fa fa-user" />
            <span> Online</span>
          </div>
        </div>
      </li>);
    });

    return result.filter(value => value);
  }

  render() {
    const list = this.renderList();

    const view = list.length
      ? <ul className="list">
        {this.renderList()}
      </ul>
      : <>does not exist User</>;

    return <div className="container">
      <div className="people-list" id="people-list">
        <div className="users">
          {view}
        </div>
      </div>
    </div>;
  }
}

const mapDispatchToProps = {
};

const mapStateToProps = (state: StoreState) => {
  const { room, user } = state;
  return {
    room: room.room,
    user: user.user,
    roomId: room.roomId,
    users: user.users,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(InviteUser);
