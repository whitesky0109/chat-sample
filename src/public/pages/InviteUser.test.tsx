import React from 'react';
import { Provider } from 'react-redux';
import { mount, ReactWrapper } from 'enzyme';
import createMockStore from 'redux-mock-store';

import InviteUser, { Props } from './InviteUser';

import { StoreState } from 'models/client';

describe('InviteUser', () => {
  let component: ReactWrapper;
  let inviteComp: ReactWrapper;
  const mock: any = jest.fn();
  const props: Props = {
    match: mock,
    location: mock,
    history: {
      goBack: mock,
    } as any,
  };

  const generateFakeStore = (initState: StoreState) => {
    const mockStore = createMockStore();
    return mockStore(initState);
  };
  describe('기본 동작', () => {
    it('render', () => {
      const store = generateFakeStore({
        user: {
          user: 'userid',
          users: {
            ['userid']: 'sockid1',
          },
        },
        room: {
          roomId: 'roomid',
          room: {
            ['roomid']: {
              creator: 'userid',
              name: 'roomname',
              users: {
                ['userid']: 'sockid1',
              },
            },
          },
        },
      });
      component = mount(<Provider store={store}>
        <InviteUser {...props} />
      </Provider>);
    });

    it('matches snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('find InviteUser component', () => {
      inviteComp = component.find('InviteUser');
      expect(inviteComp).toHaveLength(1);
    });

    it('초대할 대상이 없는 메세지 확인', () => {
      const text = inviteComp.find('.users').text();
      expect(text).toBe('초대할 수 있는 사용자가 없습니다.');
    });

    it('채팅방 이름 확인', () => {
      const text = inviteComp.find('.chat-num-messages').text();
      expect(text).toBe('Room Name : roomname');
    });

    it('취소 버튼 확인', () => {
      inviteComp.find('.btn').simulate('click');
      expect(mock).toHaveBeenCalled();
    });
  });

  describe('사용자 선택', () => {
    it('render', () => {
      const store = generateFakeStore({
        user: {
          user: 'userid',
          users: {
            ['userid']: 'sockid1',
            ['userid2']: null,
            ['userid3']: 'sockid2',
          },
        },
        room: {
          roomId: 'roomid',
          room: {
            ['roomid']: {
              creator: 'userid',
              name: 'roomname',
              users: {
                ['userid']: 'sockid1',
              },
            },
          },
        },
      });
      component = mount(<Provider store={store}>
        <InviteUser {...props} />
      </Provider>);
    });

    it('find InviteUser component', () => {
      inviteComp = component.find('InviteUser');
      expect(inviteComp).toHaveLength(1);
    });

    it('초대 가능 사용자 수 확인', () => {
      const liList = inviteComp.find('li');
      expect(liList.length).toBe(2);
    });

    it('오프라인 사용자 확인', () => {
      const li = inviteComp.find('li').at(0);
      const status = li.find('.fa-circle').prop('className');

      expect(status).not.toBeUndefined();
      if (status) {
        expect(status.split(' ')[2]).toBe('offline');
      }
    });

    it('온라인 사용자 확인', () => {
      const li = inviteComp.find('li').at(1);
      const status = li.find('.fa-circle').prop('className');

      expect(status).not.toBeUndefined();
      if (status) {
        expect(status.split(' ')[2]).toBe('online');
      }
    });

    it('사용자 선택 확인', () => {
      const li = inviteComp.find('li').at(0);
      li.simulate('click');
      expect(mock).toHaveBeenCalled();
    });

  });

});
