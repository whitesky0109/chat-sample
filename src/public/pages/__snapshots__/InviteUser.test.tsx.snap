// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InviteUser 기본 동작 matches snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(InviteUser)
    history={
      Object {
        "goBack": [MockFunction],
      }
    }
    location={[MockFunction]}
    match={[MockFunction]}
  >
    <InviteUser
      history={
        Object {
          "goBack": [MockFunction],
        }
      }
      location={[MockFunction]}
      match={[MockFunction]}
      room={
        Object {
          "roomid": Object {
            "creator": "userid",
            "name": "roomname",
            "users": Object {
              "userid": "sockid1",
            },
          },
        }
      }
      roomId="roomid"
      user="userid"
      users={
        Object {
          "userid": "sockid1",
        }
      }
    >
      <div
        className="container"
      >
        <div
          className="header"
        >
          <div
            className="title"
          >
            <div
              className="user-icon"
            >
              <i
                className="fa fa fa-user"
              />
            </div>
            <div
              className="about"
            >
              <div
                className="chat-with"
              >
                Invite User
              </div>
              <div
                className="chat-num-messages"
              >
                Room Name : 
                roomname
              </div>
            </div>
          </div>
          <div
            className="btn-group"
          >
            <button
              className="btn"
              onClick={[Function]}
              title="취소"
            >
              <i
                className="fa fa-share-square-o"
              />
            </button>
          </div>
        </div>
        <div
          className="people-list"
          id="people-list"
        >
          <div
            className="users"
          >
            초대할 수 있는 사용자가 없습니다.
          </div>
        </div>
      </div>
    </InviteUser>
  </Connect(InviteUser)>
</Provider>
`;
